File: set1/6/main.cpp
#include<iostream>
#include<climits>

using namespace std;

int arr[1003][1003];
int DP[2][1003][1003];

int main(){
    ios::sync_with_stdio(false);
    cin.tie(0);

    int N, M; cin >> N >> M;
    for(int i = 1; i <= N; ++i) {
        for(int j = 1; j <= M; ++j) {
            cin >> arr[i][j];
        }
    }
    for(int i = 0; i <= N + 1; ++i) {
        for(int j = 0; j <= M + 1; ++j) {
            DP[0][i][j] = DP[1][i][j] = INT_MIN / 2;
        }
    }

    DP[0][N][1] = arr[N][1];
    DP[1][N][M] = arr[N][M];
    for(int i = N; i >= 1; --i) {
        for(int j = 1; j <= M; ++j) {
            if(i == N && j == 1) continue;
            DP[0][i][j] = max(DP[0][i + 1][j], DP[0][i][j - 1]) + arr[i][j];
        }
    }
    for(int i = N; i >= 1; --i) {
        for(int j = M; j >= 1; --j) {
            if(i == N && j == M) continue;
            DP[1][i][j] = max(DP[1][i + 1][j], DP[1][i][j + 1]) + arr[i][j];
        }
    }

    int ans = INT_MIN / 2;
    for(int i = 1; i <= N; ++i) {
        for(int j = 1; j <= M; ++j) {
            ans = max(ans, DP[0][i][j] + DP[1][i][j]);
        }
    }
    cout << ans;

    return 0;
}

File: set1/6/main.py
import sys

def input():
    return sys.stdin.readline().rstrip()

INT_MIN = -2 ** 30
N, M = map(int, input().split())
arr = [[0 for j in range(M + 3)] for i in range(N + 3)]
DP = [[[INT_MIN for j in range(M + 3)] for i in range(N + 3)] for k in range(2)]

for i in range(1, N + 1): arr[i][1:N+1]=list(map(int,input().split()))

DP[0][N][1], DP[1][N][M] = arr[N][1], arr[N][M]
for i in range(N, 0, -1):
    for j in range(1, M + 1, 1):
        if i == N and j == 1: continue
        DP[0][i][j] = max(DP[0][i + 1][j], DP[0][i][j - 1]) + arr[i][j]

for i in range(N, 0, -1):
    for j in range(M, 0, -1):
        if i == N and j == M: continue
        DP[1][i][j] = max(DP[1][i + 1][j], DP[1][i][j + 1]) + arr[i][j]

ans = INT_MIN
for i in range(1, N + 1):
    for j in range(1, M + 1):
        ans = max(ans, DP[0][i][j] + DP[1][i][j])
print(ans)